lib/src/rules/prefer_enums.dart

```diff
@@ -0,0 +1,254 @@
+        throw _InvalidEnumException();
+      } else if (element.interfaces
+          .map((e) => e.element)
+          .contains(classElement)) {
+        throw _InvalidEnumException();
+      } else if (element.mixins.map((e) => e.element).contains(classElement)) {
+        // This case won't occur unless there's an error in the source code, but
+        // it's easier to check for the condition than it is to check for the
+        // diagnostic.
+        throw _InvalidEnumException();
+      }
+      super.visitClassDeclaration(node);
+    }
+  }
+
+  @override
+  void visitInstanceCreationExpression(InstanceCreationExpression node) {
+    if (invokesGenerativeConstructor(node)) {
+      throw _InvalidEnumException();
+    }
+    super.visitInstanceCreationExpression(node);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  final LinterContext context;
+
+  _Visitor(this.rule, this.context);
+
+  @override
+  visitClassDeclaration(ClassDeclaration node) {
+    if (node.isAbstract) return;
+    var classElement = node.declaredElement;
+    if (classElement == null) return;
+
+    // Enums can only extend Object.
+    if (node.extendsClause?.superclass.type?.isDartCoreObject == false) {
+      return;
+    }
+
+    var candidateConstants = <VariableDeclaration>[];
+
+    for (var member in node.members) {
+      if (isHashCode(member)) return;
+      if (isIndex(member)) return;
+      if (isEquals(member)) return;
+
+      if (member is FieldDeclaration) {
+        if (!member.isStatic) continue;
+        for (var field in member.fields.variables) {
+          var fieldElement = field.declaredElement;
+          if (fieldElement is! FieldElement) continue;
+          if (field.isSynthetic || !field.isConst) continue;
+          var initializer = field.initializer;
+          if (initializer is! InstanceCreationExpression) continue;
+
+          var constructorElement = initializer.constructorName.staticElement;
+          if (constructorElement == null) continue;
+          if (constructorElement.isFactory) continue;
+          if (constructorElement.enclosingElement != classElement) continue;
+
+          if (fieldElement.computeConstantValue() == null) continue;
+
+          candidateConstants.add(field);
+        }
+      }
+      if (member is ConstructorDeclaration) {
+        var constructor = member.declaredElement;
+        if (constructor == null) return;
+        if (!constructor.isFactory && !constructor.isConst) return;
+        var name = member.name?.name;
+        if (classElement.isPublic &&
+            (name == null || !Identifier.isPrivateName(name))) {
+          return;
+        }
+      }
+    }
+
+    if (candidateConstants.length < 2) return;
+    if (DartTypeUtilities.hasSubclassInDefiningCompilationUnit(classElement)) {
+      return;
+    }
+
+    try {
+      node.accept(_EnumVisitor(classElement, candidateConstants));
+    } on _InvalidEnumException {
```

Comment:
I preserved this approach from the assist but expect that it would be more performant to use a flag rather than a thrown exception to short-circuit the common negative case.

@bwilkerson: wdyt?  Should I go with my gut and refactor?